DVCodingStyle.md

```diff
@@ -454,13 +455,194 @@ utility macros for even more common code sequences.
+    //      which is not equal to ((p || q) && r) || s.
+    ```
+    :+1:
+    ```systemverilog
+    `define AND(a, b) ((a) && (b))
+
+    assign x = `AND(p || q, r) || s;
+    // Good: Output in-line with the expectation.
+    ```
+2.  Macro usage scope and avoiding namespace collisions.
+
+    Macros have a global scope regardless of where they are defined. Once
+    compiled, they are visible to ALL subsequent sources in the same
+    compilation unit as well as to all subsequent compilation units. Hence,
+    care must be taken to prevent macro re-definitions. Some tools do not even
+    warn about name collisions, causing unnecessary debug overhead.
+
+    *   Macro names must be appropriately prefixed with a thematically chosen
+        string. It is typical to use the file name or the package name as the
+        prefix string for all macros defined within that source.
+
+        :-1:
+        ```systemverilog
+        // file: dv_macros.h
+        `define STRINGIFY(s) `"s`"
+        // Bad: UVM library code also defines a macro with the same name.
+        //      Depending on the tool, this may be flagged as an error.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: dv_macros.h
+        `define DV_STRINGIFY(s) `"s`"
+        ```
+
+    *   If the scope of the macros defined are global in nature (useful for
+        the entire testbench), then place them in a separate SystemVerilog
+        header file with `` `ifndef`` guards. It must only contain macro
+        definitions and nothing else. The file name must be suffixed with
+        `_macros` and have the `.svh` extension to denote that it is a header
+        file.
+
+        :-1:
+        ```systemverilog
+        // file: dv_macros.h
+        automatic function void get_nco();
+          ...
+        endfunction
+        // Bad: Putting functions and tasks in a header file that is potentially
+        //      included multiple times.
+
+        `define DV_FOO(...) ...
+        // Bad: No ifndef guard. Macro re-definition warnings will be thrown if
+        //      this header file is included in multiple sources.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: dv_macros.svh
+        `ifndef DV_FOO
+          `define DV_FOO(...) ...
+        `endif
+
+        `ifndef DV_BAR
+          `define DV_BAR xyz
+        `endif
+        ```
+
+        Generally, there is no need to `` `include`` this header file if it
+        can be guaranteed that it is already compiled prior to the current
+        compilation unit in which those macros are invoked. This is typically
+        handled by an external tool that works out the dependencies across
+        the compilation units (example: FuseSoC). However, it may not always be
+        possible to guarantee this, and the `` `include`` may be necessary.
+        The `` `ifndef`` guard helps prevent the re-definition of the macros with
+        itself.
```

Comment:
I agree that this statement is true, but I'm not sure it's good style. I'm in favour of trying to remember to include all headers that you use, and relying on include guards to avoid duplication. Otherwise, you get weird problems when you change the includes in one file, and a completely different one suddenly breaks.

Reply:
I'm not sure this matches how we currently use files like `prim_assert.svh`. Also, note that this style guide doesn't actually mention FuseSoc, so I don't think it makes sense to talk about how FuseSoc works (which I suspect might be tool-dependent anyway) here.

@imphil: I know you've fixed a couple of missing includes like this. What do you think?

Reply:
Whatever build system you use to manage your SV sources, it will ultimately generate a dependency - resolved file list (the .f file is kind of an industry standard that all ASIC tool flows use) that is fed to the ASIC tool. FuseSoC is provided as an example - all ASIC build systems will work the same way.

Reply:
The big caveat is "if it can be guaranteed that it is already compiled prior to the current compilation unit". We neither require nor guarantee nor implicitly fulfill this behavior at the moment for all EDA tools.

Background: Most EDA tools can be used in a way that generates a single compilation unit for the whole project, or in a way that generates a single compilation unit per .sv source file (i.e. how C/C++ compilation is typically done). (And typically arbitrary in-between approaches.)
Edalize, at the moment, does not guarantee either behavior. Even more, the different backends implement things differently: the VCS and Vivado backends, for example, put everything into a single compilation unit. The Aldec Riviera-PRO backend, on the other hand, creates individual compilation units.


Personal comment: I prefer including all dependencies in each file. This is a common approach in most programming languages, all preferring explicit coding over implicit behavior. I'd therefore remove the whole statement from the style guide, and possibly even turn it around to require all dependencies to be included.

Reply:
@imphil, regardless if whether you provide one monolithic compilation unit to the simulator, or individual one for each package, unlike C, macros definitions once compiled, are visible to all sources / compilation units without needing to `` `include`` header files repeatedly. This is because in SV, a header file is yet another source like any other SV source.

Like I said in my previous comment - if you add the header source as a dependency in the FuseSoC core file, that should itself guarantee that the macro header is compiled first. I see the problem with `prim_assert` - in its core file, it has the tag **`is_include_file: true`**, which does not place it in the file list for compilation immediately, but only adds its  search path (`+incdir+<path>`). This forces the user to not only `` `include`` it in all sources, but also list it as a FuseSoC core dependency (so that the search path is added to all "compilation units"). It also forces the addition of `ifndef` guards to prevent its own recompilation! So I'm really not sure how is the extra line `` `include prim_assert.svh`` added to every single source even useful. If you remove the `is_include_file: true` tag, all you need is the FuseSoC core dependency (which IMO is already explicit enough). 

In any case, the current text already recommends adding the `` `ifndef`` protection against macro-redefinitions with itself due to the header `` `included`` multiple times. I will reword it a bit. 

Considering UVM macros are used ALL over the place without actually including `uvm_macros.svh`, I will not be adding a recommendation to explicitly include the macro header files in every single source.

Reply:
I have removed the controversial line altogether. PTAL, thanks!